Source/DafnyCore/Compilers/Compiler-go.cs

```diff
@@ -2061,12 +2061,36 @@ private class SpecialNativeType : UserDefinedType {
         wr.Write((bool)e.Value ? "true" : "false");
       } else if (e is CharLiteralExpr) {
         var v = (string)e.Value;
-        wr.Write("_dafny.Char('{0}')", TranslateEscapes(v, isChar: true));
+        wr.Write("_dafny.Char(");
+        // See comment on the StringLiteralExpr case below.
+        if (Util.Utf16Escape.IsMatch(v)) {
+          char c = Util.UnescapedCharacters(v, false).Single();
+          wr.Write($"{(int)c}");
+        } else {
+          wr.Write("'{0}'", TranslateEscapes(v, isChar: true));
+        }
+        wr.Write(")");
       } else if (e is StringLiteralExpr) {
         var str = (StringLiteralExpr)e;
-        wr.Write("_dafny.SeqOfString(");
-        TrStringLiteral(str, wr);
-        wr.Write(")");
+        // It may not be possible to translate \u escapes into a valid Go string,
+        // since Go string literals have to be encodable in UTF-8,
+        // but Dafny allows invalid sequences of surrogate characters.
+        // So if any are present, just emit a sequence of the direct UTF-16 code units instead.
+        var s = (string)str.Value;
+        if (!str.IsVerbatim && Util.Utf16Escape.IsMatch(s)) {
```

Comment:
What happens if the C# string contains unpaired surrogates (not `\u` escaped, just directly, unescaped, in the string?)  Is that impossible thanks to the way we parse?

Reply:
It SHOULD be impossible because Dafny source files have to be UTF-8 encoded, and the scanner is decoding to `int` code points: https://github.com/dafny-lang/dafny/blob/master/Source/DafnyCore/Coco/Scanner.frame#L210

That code doesn't look particularly robust in the face of invalid UTF-8 sequences mind you, but it doesn't look like it's possible to produce surrogate values.

There's also the fact that the generated scanner code includes a straight cast from `int` to `char`, which could truncate a code point to `0xFFFF`. I've got a fix for that in the unicode char branch (and I should probably cut an issue for it right away) but that still won't create surrogates at least.

Reply:
OK, sweet.  I'll let you judge whether it's worth a Contract.Assert here — it might save our bacon down the line?

Reply:
Actually there's nothing to assert here: I'm specifically NOT assuming surrogates are used correctly here. The only concern is accurately translating escape sequences to target language escape sequences (or something else equivalent). I'm not aware of any bugs in translating arbitrary `char` values in the string at least.